### PR TITLE
Multi-gpu support in training loop

### PR DESCRIPTION
**Pull request by ehsanmok**
*Issue #162 

*Description of changes:*

Adding the multi-gpu support. Besides local test, **all** examples are replicable and been tested on p3.8xlarge with 4 gpus.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/model/test_models_synthetic.py [hunk 1]
```diff
@@ -32,13 +33,18 @@
 batch_size = 32
 context_length = 2
 epochs = 1
+ctx = (
```
Comment (jaheba) [inline test/model/test_models_synthetic.py]: Couldn't we use `normalize_ctx` here?

File: src/gluonts/core/component.py [hunk 2]
```diff
@@ -506,6 +508,21 @@ def get_mxnet_context(gpu_number=0) -> mx.Context:
...
+) -> List[mx.Context]:
+    if isinstance(ctx, str):
+        return [mx.Context(ctx)]
+    if ctx is not None:
+        return [ctx] if not isinstance(ctx, list) else ctx
+    else:
+        return (
+            [mx.gpu(i) for i in range(mx.context.num_gpus())]
+            if mx.context.num_gpus()
+            else [mx.cpu()]
```
Comment (jaheba) [inline src/gluonts/core/component.py]: I think I prefer this:

```python
def normalize_ctx(
    ctx: Optional[Union[str, mx.Context, List[mx.Context]]]
) -> List[mx.Context]:
    if ctx is None:
        if mx.context.num_gpus():
            return [mx.gpu(i) for i in range(mx.context.num_gpus())]
        else:
            return [mx.cpu()]

    if isinstance(ctx, str):
        return [mx.Context(ctx)]

    if isinstance(ctx, list):
        return ctx

    return [ctx]
```

File: src/gluonts/core/component.py [hunk 3]
```diff
@@ -506,6 +508,21 @@ def get_mxnet_context(gpu_number=0) -> mx.Context:
         return mx.context.cpu()
 
 
+def normalize_ctx(
+    ctx: Optional[Union[str, mx.Context, List[mx.Context]]]
```
Comment (jaheba) [inline src/gluonts/core/component.py]:
```suggestion
    ctx: Union[str, ContextType]
```

File: src/gluonts/dataset/loader.py [hunk 4]
```diff
@@ -14,19 +14,14 @@
 # Standard library imports
 import functools
 import itertools
-from typing import (
-    Any,
-    Dict,
-    Iterable,
-    Iterator,
-)
+from typing import Any, Dict, Iterable, Iterator, List, Union
```
Comment (jaheba) [inline src/gluonts/dataset/loader.py]: Are you using `List` and `Union`?

File: src/gluonts/dataset/loader.py [hunk 5]
```diff
@@ -79,7 +74,11 @@ def stack(self, xs):
             data = np.asarray(xs)
             if data.dtype.kind == "f":
                 data = data.astype(self.dtype)
-            return mx.nd.array(data, dtype=data.dtype, ctx=self.ctx)
+            return mx.nd.array(
+                data,
+                dtype=data.dtype,
+                ctx=self.ctx[0] if isinstance(self.ctx, list) else self.ctx,
```
Comment (jaheba) [inline src/gluonts/dataset/loader.py]: Shouldn't this just be `ctx[0]`?

File: test/model/test_models_synthetic.py (hunk 1 above)
Comment (ehsanmok) [reply test/model/test_models_synthetic.py]: Can make it to `ctx=normalize_ctx(None)` instead but it's less clear I guess, so being explicit here.

File: src/gluonts/core/component.py (hunk 2 above)
Comment (ehsanmok) [reply src/gluonts/core/component.py]: No sure the difference! would it make it more readable?

Comment (jaheba) [reply src/gluonts/core/component.py]: Yes, handling `None` first, then the other cases and all in one level.

File: src/gluonts/core/component.py (hunk 3 above)
Comment (ehsanmok) [reply src/gluonts/core/component.py]: The idea is to keep the `Optional` and `str` in `ContextType` to make it compatible with mxnet.

Comment (jaheba) [reply src/gluonts/core/component.py]: Since `ContextType` already includes `Optional`, we can shorten this.

File: src/gluonts/dataset/loader.py (hunk 4 above)
Comment (ehsanmok) [reply src/gluonts/dataset/loader.py]: Thanks! removing them. Was before the `ContextType` refactoring.

**Comment by ehsanmok** [comment]
@jaheba thanks for the review! all the comments have been addressed. I avoid making incompatible names changes to mxnet so keeping the `ctx` name.

**Comment by alexw91** [comment]
# [Codecov](https://codecov.io/gh/awslabs/gluon-ts/pull/657?src=pr&el=h1) Report
> Merging [#657](https://codecov.io/gh/awslabs/gluon-ts/pull/657?src=pr&el=desc) into [master](https://codecov.io/gh/awslabs/gluon-ts/commit/f6c12a4a3902a6f2137a0f6b98001c64b9cdb9a7&el=desc) will **increase** coverage by `0.00%`.
> The diff coverage is `n/a`.


[Impacted file tree graph](https://codecov.io/gh/awslabs/gluon-ts/pull/657?src=pr&el=tree)

```diff
@@           Coverage Diff           @@
##           master     #657   +/-   ##
=======================================
  Coverage   84.63%   84.64%           
=======================================
  Files         178      178           
  Lines       10401    10425   +24     
=======================================
+ Hits         8803     8824   +21     
- Misses       1598     1601    +3     
```

| [Impacted Files](https://codecov.io/gh/awslabs/gluon-ts/pull/657?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/gluonts/core/component.py](https://codecov.io/gh/awslabs/gluon-ts/pull/657/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvY29yZS9jb21wb25lbnQucHk=) | `86.22% <0.00%> (-0.96%)` | :arrow_down: |

**Comment by alexw91** [comment]
# [Codecov](https://codecov.io/gh/awslabs/gluon-ts/pull/657?src=pr&el=h1) Report
> Merging [#657](https://codecov.io/gh/awslabs/gluon-ts/pull/657?src=pr&el=desc) into [master](https://codecov.io/gh/awslabs/gluon-ts/commit/f6c12a4a3902a6f2137a0f6b98001c64b9cdb9a7&el=desc) will **increase** coverage by `0.00%`.
> The diff coverage is `n/a`.


[Impacted file tree graph](https://codecov.io/gh/awslabs/gluon-ts/pull/657?src=pr&el=tree)

```diff
@@           Coverage Diff           @@
##           master     #657   +/-   ##
=======================================
  Coverage   84.63%   84.64%           
=======================================
  Files         178      178           
  Lines       10401    10425   +24     
=======================================
+ Hits         8803     8824   +21     
- Misses       1598     1601    +3     
```

| [Impacted Files](https://codecov.io/gh/awslabs/gluon-ts/pull/657?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/gluonts/core/component.py](https://codecov.io/gh/awslabs/gluon-ts/pull/657/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvY29yZS9jb21wb25lbnQucHk=) | `86.22% <0.00%> (-0.96%)` | :arrow_down: |

**Comment by istorch** [comment]
@ehsanmok @jaheba Sorry to just randomly comment here (this is my first comment on this project), but when can we expect to see this released? I am a new user of GluonTS and it would be nice to have multi GPU support (my specific application uses DeepAR). Thanks!

**Comment by lostella** [comment]
@istorch the changes look good to me, so this could be merged into master soon and be included in the next release (either this or next week)

File: src/gluonts/dataset/loader.py [hunk 6]
```diff
@@ -96,7 +96,7 @@ def __init__(
             cyclic=self.cyclic,
             is_train=self.is_train,
             batch_size=self.batch_size,
-            ctx=self.ctx,
+            ctx=self.ctx[0],
```
Comment (lostella) [inline src/gluonts/dataset/loader.py:99]: I'm not sure I understand this: one can specify a list of contexts, but then the data is loaded in the first one? But I'm really ignorant about the argument, so maybe I'm missing something.

File: src/gluonts/model/deepar/_network.py [hunk 7]
```diff
@@ -266,19 +266,27 @@ def unroll_encoder(
...
+                else inputs[0].context
+            )
+            with ctx:
+                begin_state = self.rnn.begin_state(
+                    func=F.zeros, dtype=self.dtype, batch_size=inputs.shape[0]
+                )
+        else:
+            begin_state = self.rnn.begin_state(
+                func=F.zeros, dtype=self.dtype, batch_size=0
+            )
```
Comment (lostella) [inline src/gluonts/model/deepar/_network.py:282]: It looks to me like this context logic should not be in the network code. Maybe it would be better to control the context once and for all from the outside somehow? Otherwise I can see how the code of every model could be filled with similar context logic, and lose readability. What do you think?

**Comment by istorch** [comment]
@ehsanmok @lostella Sorry to bug you guys again, but has there been any progress on this PR?

File: src/gluonts/support/util.py [hunk 8]
```diff
@@ -158,9 +160,16 @@ def copy_parameters(
...
+        for p in params:
+            ctx = params[p].list_ctx()
+            break
+        # force init otherwise may not load params
+        # since not all params have the same ctx in net_source
+        net_dest.initialize(ctx=ctx, force_reinit=True)
         net_dest.load_parameters(
             model_dir_path,
-            ctx=mx.current_context(),
+            ctx=ctx,
```
[the remark this answers is not in the document]
Comment (ehsanmok) [reply src/gluonts/support/util.py:172]: If I remember, during testing, when not all params are in a single context, forcing to reinit with a list of contexts solved the issue. Please see my general comment at the end.

File: src/gluonts/model/deepar/_network.py (hunk 7 above)
Comment (ehsanmok) [reply src/gluonts/model/deepar/_network.py:282]: Agree and in fact this has been repeated. Though I was expecting to have this in upstream mxnet to not hack this into.

**Comment by ehsanmok** [comment]
@lostella Thanks for your comments! in general, I think, since gluonts hasn't taken multi-gpu into considerations and has been designed for single gpu (it felt like that at least at the time of creating this PR), this PR initiations a sub-optimal solution at least with rooms to improvements of course. 

As for the MVE, I've run all the existing examples and observed their improved sub-optimal speed ups which I can report later here.

**Comment by ehsanmok** [comment]
Hi @istorch, thanks for your patience! As @lostella pointed out earlier, this PR offers a solution and it'd be helpful to have some external validations too, so if you can, please run your application to test things us further. You can install this patch via 

```
python -m pip install git+git://github.com/ehsanmok/gluon-ts@multi-gpu
```

File: src/gluonts/dataset/loader.py (hunk 6 above)
Comment (ehsanmok) [reply src/gluonts/dataset/loader.py:99]: > is the idea to put everything in the first context, and then have split_and_load in the training loop take care of that?

Yes, indeed!

**Comment by istorch** [comment]
@ehsanmok Thanks for picking this up again! I have installed your branch and passed the context in as a list, like `ctx = [mx.gpu(i) for i in range(8)]`. Watching nvidia-smi, I see that all 8 GPUs are utilized during training. However, when I go to generate predictions, it still only uses one GPU. Is there anything special I need to do to make it use all GPUs for prediction as well?

**Comment by ehsanmok** [comment]
@istorch thanks a lot for the report! I'll look into the prediction part.

**Comment by istorch** [comment]
@ehsanmok also, I did a little benchmarking and I'm not seeing any improvement in using multiple GPUs. With the same settings, my training ran in 7min 40s on a single GPU, and 8min 51s on 8 GPUs. I set epochs to 50, batch_size to 512, and num_batches_per_epoch to 16 (this is on a p2.8xlarge instance).

**Comment by ehsanmok** [comment]
@istorch Just curious, did you multiply the batch size by the number of GPUs, like if you've used 512 for single GPU, then you'd need to use 512 * 8 for your 8 GPU context? Also it'd be helpful if you could share more context about what your code is doing and what model are you training.

**Comment by ehsanmok** [comment]
@lostella @istorch after an investigation it turns out the adding the complete multi-gpu feature is harder than what I initially thought. This PR add the multi-gpu for training loop only. The current data loader seems to be a major bottleneck as it wasn't written for single context only and the other parts of GluonTS estimator is hard to decouple and extend. This is one more reason to consider #829.
